smart_open/compression.py

```diff
@@ -62,20 +62,22 @@ def _handle_gzip(file_obj, mode):
     return gzip.GzipFile(fileobj=file_obj, mode=mode)
 
 
-def compression_wrapper(file_obj, mode):
+def compression_wrapper(file_obj, mode, filename=None):
     """
     This function will wrap the file_obj with an appropriate
     [de]compression mechanism based on the extension of the filename.
 
     file_obj must either be a filehandle object, or a class which behaves
-        like one.  It must have a .name attribute.
+    like one. It must have a .name attribute unless `filename` is given.
 
     If the filename extension isn't recognized, will simply return the original
     file_obj.
-    """
 
+    """
     try:
-        _, ext = os.path.splitext(file_obj.name)
+        if filename is None:
```

Comment:
> The conditional can never raise any of the below exceptions

The conditional can raise exceptions. That's why I moved it inside try-catch in https://github.com/RaRe-Technologies/smart_open/pull/497/commits/9d6ff37492d7516e3644dee5b9b49d315170c411

> unable to transparently decompress %r
because .name attribute is missing (consider passing
filename as a keyword argument).

That would only confuse users IMO. They have no option of "passing filename as a keyword argument" – this function is used internally by smart_open. Users have control over what is passed at all.

The only way they can pass `filename` is if they somehow find this function and call it explicitly, manually. In which case the message is not helpful anyway.